src/main/java/io/cdap/plugin/MultiTableDBSource.java

```diff
@@ -73,19 +87,72 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
   public void prepareRun(BatchSourceContext context) throws Exception {
     Configuration hConf = new Configuration();
     Class<? extends Driver> driverClass = context.loadPluginClass(JDBC_PLUGIN_ID);
-    Collection<DBTableInfo> tables = MultiTableDBInputFormat.setInput(hConf, conf, driverClass);
-    SettableArguments arguments = context.getArguments();
-    for (DBTableInfo tableInfo : tables) {
-      arguments.set(DynamicMultiFilesetSink.TABLE_PREFIX + tableInfo.getDbTableName().getTable(),
-                    tableInfo.getSchema().toString());
-    }
 
-    context.setInput(Input.of(conf.getReferenceName(),
-                              new SourceInputFormatProvider(MultiTableDBInputFormat.class, hConf)));
+    Collection<DBTableInfo> tables;
+
+    try {
+      tables = MultiTableDBInputFormat.setInput(hConf, conf, driverClass);
+
+      SettableArguments arguments = context.getArguments();
+      for (DBTableInfo tableInfo : tables) {
+        arguments.set(DynamicMultiFilesetSink.TABLE_PREFIX + tableInfo.getDbTableName().getTable(),
+                      tableInfo.getSchema().toString());
+      }
+
+      context.setInput(Input.of(conf.getReferenceName(),
+                                new SourceInputFormatProvider(ErrorCollectingMultiTableDBInputFormat.class, hConf)));
+    } catch (Exception ex) {
+      String errorMessage = "Error getting table schemas from database.";
+      LOG.error(errorMessage, ex);
+
+      InputFormatProvider errorEmittingInputFormatProvider = new InputFormatProvider() {
+        @Override
+        public String getInputFormatClassName() {
+          return ErrorEmittingInputFormat.class.getName();
+        }
+
+        @Override
+        public Map<String, String> getInputFormatConfiguration() {
+          Map<String, String> config = new HashMap<>();
+          for (Map.Entry<String, String> entry : hConf) {
+            config.put(entry.getKey(), entry.getValue());
+          }
+          //Add ErrorEmittingInputFormat config properties
+          config.put(ErrorEmittingInputFormat.ERROR_MESSAGE, errorMessage);
+          config.put(ErrorEmittingInputFormat.EXCEPTION_CLASS_NAME, ex.getClass().getCanonicalName());
+
+          return config;
+        }
+      };
+
+      context.setInput(Input.of(conf.getReferenceName(), errorEmittingInputFormatProvider));
+    }
   }
 
   @Override
   public void transform(KeyValue<NullWritable, StructuredRecord> input, Emitter<StructuredRecord> emitter) {
-    emitter.emit(input.getValue());
+    StructuredRecord record = input.getValue();
+
+    // Check if the record is an error.
+    if (ErrorSchema.SCHEMA_NAME.equals(record.getSchema().getRecordName())) {
+      numErrors++;
+
+      // Fail the pipeline if the table error threshold exceeds the desired number of table failures.
+      if (MultiTableConf.ERROR_HANDLING_FAIL_PIPELINE.equals(conf.getErrorHandlingMode()) &&
```

Comment:
This logic here does not really as the pipeline is split across multiple workers. I need to find a way to keep this global counter shared across workers.

Reply:
there isn't a way to make this global, it has to be per split.

Reply:
the solution today for global error counts is to use a condition plugin in the middle of the pipeline, which has access to metrics from all the previous stages.

Reply:
I see. I'll communicate this on the ticket.